tests/test_array_compat.py

```diff
@@ -0,0 +1,140 @@
+"""Test array libraries compatibility utils."""
+
+import pytest
+
+from pytest_cases import parametrize, fixture
+from unittest.mock import patch, MagicMock
+
+import numpy as np
+
+from mrinufft._array_compat import (
+    with_numpy,
+    with_numpy_cupy,
+    with_torch,
+    with_tensorflow,
+    _get_device,
+    CUPY_AVAILABLE,
+    AUTOGRAD_AVAILABLE,
+    TENSORFLOW_AVAILABLE,
+    tf_cuda_is_available,
+)
+
+from helpers import to_interface
+from helpers.factories import _param_array_interface
+
+if CUPY_AVAILABLE:
+    import cupy as cp
+
+torch_cuda_is_available = False
+if AUTOGRAD_AVAILABLE:
+    import torch
+
+    torch_cuda_is_available = torch.cuda.is_available()
+
+if TENSORFLOW_AVAILABLE:
+    import tensorflow as tf
+
+
+def dummy_func(*args):
+    return args
+
+
+@fixture(scope="module")
+@parametrize("decorator", [with_numpy, with_numpy_cupy, with_torch, with_tensorflow])
+def decorator_factory(request, decorator):
+    @decorator
+    def test_func(*args):
+        return dummy_func(*args)
+
+    return test_func, decorator
+
+
+@_param_array_interface
+def test_decorators_outcome(decorator_factory, array_interface):
+    decorated_function, _ = decorator_factory
+
+    # Create input array
+    array = to_interface(np.asarray([1.0, 2.0, 3.0]), array_interface)
+
+    # Output device and type must be the same as leading argument
+    expected_type = type(array)
+    expected_device = _get_device(array)
+
+    # Execute function
+    outputs = decorated_function(array, array, array)
+
+    # Assert the output has correct type
+    for output in outputs:
+        assert isinstance(
+            output, expected_type
+        ), f"Expected {expected_type} but got {type(output)}"
+        assert (
+            _get_device(output) == expected_device
+        ), f"Expected {expected_device} but got {_get_device(output)}"
+
+
+@_param_array_interface
+def test_internal_conversions(decorator_factory, array_interface):
```

Comment:
I like this,  but I think it could be refactor/parametrize  properly. An attempt:

```python
@paramtrize(with_decorator=[with_numpy,  ... ]) # probably needs some skips as well 
@parametrize("input_array_interface, array_fun") = [("numpy", "numpy.array"), ("cupy","cupy.array"), ...])
def test_internal_conversion(with_decorator, input_array_interface, array_fun):

    # Create input array
    array = to_interface(np.asarray([1.0, 2.0, 3.0]), array_interface)
    
    # Execute function and monitor internal conversion
    with patch(array_fun, wraps=eval(array_fun)) as mock_fun:
        _  = with_decorator(dummy_fun)(array, array, array, 42, "dummy")
        assert mock_fun.call_count == 3

```

An other solution would be write a test for each specific case (at least it would be more explicit which case failed in the testing)